File: packages/flutter/lib/src/logic/detail.dart
/// 详情页的纯逻辑（对应 packages/common/src/logic/detail.ts，astra.md 的 B12）。
///
/// 详情页看着只是「把一条记录摊开」，真正难的是四件事：
///
/// 一、这条记录现在能做什么。状态不允许的动作**不出现**（它在这个状态下根本
/// 不是一个动作，灰着只会让人反复去试）；没权限的动作**出现但停用并说明原因**
/// （藏起来的话用户会以为功能不存在，转头去提工单）。
///
/// 二、看到的这一份还作不作数。详情页停留期间别人改了、删了，写动作要停掉，
/// 并明说「你看到的是 v3，现在已经是 v5」——只说「操作失败」等于让人再点一次。
///
/// 三、回到列表要回到原来那个位置。
///
/// 四、上一条 / 下一条。到头了不要把按钮藏掉，说清「已经是第一条」。
library;

import 'dart:convert';

/* ---------- 这一份还作不作数 ---------- */

enum IRecordFreshness { fresh, stale, deleted }

enum IFreshnessAction { none, refresh, back }

class IFreshnessState {
  const IFreshnessState({
    required this.kind,
    required this.label,
    required this.detail,
    required this.action,
  });

  final IRecordFreshness kind;

  /// 状态本身。颜色不是唯一线索，这句话必须出现
  final String label;
  final String detail;
  final IFreshnessAction action;
}

IFreshnessState recordFreshness({
  required int seenRevision,
  required int currentRevision,
  bool exists = true,
}) {
  if (!exists) {
    return const IFreshnessState(
      kind: IRecordFreshness.deleted,
      label: '已被删除',
      detail: '这条记录已经不在了，页面上显示的是你进来时的那一份',
      action: IFreshnessAction.back,
    );
  }
  if (currentRevision > seenRevision) {
    return IFreshnessState(
      kind: IRecordFreshness.stale,
      label: '已被他人更新',
      // 只说「操作失败」等于让用户再点一次，所以把两个版本号都摆出来
      detail: '你看到的是 v$seenRevision，现在已经是 v$currentRevision',
      action: IFreshnessAction.refresh,
    );
  }
  return const IFreshnessState(
    kind: IRecordFreshness.fresh,
    label: '最新',
    detail: '',
    action: IFreshnessAction.none,
  );
}

/* ---------- 这条记录现在能做什么 ---------- */

enum IDetailActionKind { primary, standard, danger }

class IDetailActionSpec {
  const IDetailActionSpec({
    required this.key,
    required this.label,
    this.kind = IDetailActionKind.standard,
    this.states,
    this.permission,
    this.readOnly = false,
  });

  final String key;
  final String label;
  final IDetailActionKind kind;

  /// 这个动作在哪些状态下才算一个动作。null 表示任何状态都算。
  /// 不在其中就**不出现**——不是灰着。
  final List<String>? states;

  /// 需要的权限。缺了就出现但停用，并说明原因
  final String? permission;

  /// 只读动作（导出、打印）。记录失效时它们照样可用——
  /// 把「打印」一起停掉，只会让用户以为整页坏了。
  final bool readOnly;
}

class IDetailAction {
  const IDetailAction({
    required this.key,
    required this.label,
    required this.kind,
    required this.disabled,
    required this.reason,
  });

  final String key;
  final String label;
  final IDetailActionKind kind;
  final bool disabled;

  /// 停用原因。disabled 为 true 时一定不是空串
  final String reason;
}

/// 算出这一刻该摆出哪些动作、哪些是灰的、为什么灰。
///
/// 三条停用理由的优先级：权限 > 失效 > 业务规则。权限排第一是因为它最稳定——
/// 刷新一百次也不会变，先说它才不会让用户去做无用功。
List<IDetailAction> detailActions(
  List<IDetailActionSpec> specs, {
  required String status,
  List<String> permissions = const [],
  IRecordFreshness freshness = IRecordFreshness.fresh,
  Map<String, String> denied = const {},
}) {
  final granted = permissions.toSet();
  final out = <IDetailAction>[];

  for (final spec in specs) {
    // 状态不允许的动作根本不出现：灰着摆在那儿只会让人反复去试
    if (spec.states != null && !spec.states!.contains(status)) continue;

    var reason = '';
    if (spec.permission != null && !granted.contains(spec.permission)) {
      reason = '需要「${spec.permission}」权限';
    } else if (!spec.readOnly && freshness != IRecordFreshness.fresh) {
      reason = freshness == IRecordFreshness.deleted
          ? '这条记录已被删除'
          : '这条记录已被他人更新，请先刷新';
    } else if (denied[spec.key] != null) {
      reason = denied[spec.key]!;
    }

    out.add(IDetailAction(
      key: spec.key,
      label: spec.label,
      kind: spec.kind,
      disabled: reason.isNotEmpty,
      reason: reason,
    ));
  }
  return out;
}

/// 一个动作都没有时说什么。空白一片会让人以为页面没加载完
String noActionHint(String status) => '「$status」状态下没有可执行的操作';

/* ---------- 上一条 / 下一条 ---------- */

class INeighbours {
  const INeighbours({
    required this.index,
    required this.total,
    required this.prevId,
    required this.nextId,
    required this.position,
    required this.edgeHint,
  });

  /// 在这批 id 里排第几，从 0 起。不在其中是 -1
  final int index;
  final int total;
  final String? prevId;
  final String? nextId;

  /// 「第 3 条，共 128 条」。不在这批里时是空串
  final String position;

  /// 到头了说什么。没到头是空串
  final String edgeHint;
}

INeighbours detailNeighbours(List<String> ids, String currentId) {
  final index = ids.indexOf(currentId);
  final total = ids.length;
  if (index < 0) {
    return INeighbours(
      index: -1,
      total: total,
      prevId: null,
      nextId: null,
      position: '',
      edgeHint: '',
    );
  }
  final prevId = index > 0 ? ids[index - 1] : null;
  final nextId = index < total - 1 ? ids[index + 1] : null;
  // 到头了不把按钮藏掉——藏掉会让人以为是页面坏了
  final edgeHint = total <= 1
      ? '只有这一条'
      : prevId == null
          ? '已经是第一条'
          : nextId == null
              ? '已经是最后一条'
              : '';
  return INeighbours(
    index: index,
    total: total,
    prevId: prevId,
    nextId: nextId,
    position: '第 ${index + 1} 条，共 $total 条',
    edgeHint: edgeHint,
  );
}

/* ---------- 从哪儿来，回哪儿去 ---------- */

class IReturnTicket {
  const IReturnTicket({required this.search, required this.scrollY, this.focusId});

  /// 列表当时的查询串（含筛选与页码），原样带回去
  final String search;

  /// 列表当时滚到哪儿。取整到像素——小数在不同缩放下还原不回同一位置
  final int scrollY;

  /// 从哪一行点进来的。回去之后把它高亮一下
  final String? focusId;

  @override
  bool operator ==(Object other) =>
      other is IReturnTicket &&
      other.search == search &&
      other.scrollY == scrollY &&
      other.focusId == focusId;

  @override
  int get hashCode => Object.hash(search, scrollY, focusId);
}

/// 用 JSON 而不是自定义分隔符：查询串里本来就什么字符都可能有，
/// 自己拼分隔符迟早会被一个带 & 的筛选值劈开。
String packReturn(IReturnTicket ticket) {
  final payload = <String, Object?>{
    's': ticket.search,
    'y': ticket.scrollY < 0 ? 0 : ticket.scrollY,
  };
  if (ticket.focusId != null && ticket.focusId!.isNotEmpty) payload['f'] = ticket.focusId;
  return jsonEncode(payload);
}

/// 解不出来就返回 null，不抛错：这张票据来自 URL，用户会手改、会截断。
/// 调用方拿到 null 就退回列表顶部——不精确但绝不出错的落点。
IReturnTicket? unpackReturn(String? raw) {
  if (raw == null || raw.isEmpty) return null;
  try {
    final parsed = jsonDecode(raw);
    if (parsed is! Map) return null;
    final search = parsed['s'] is String ? parsed['s'] as String : '';
    final rawY = parsed['y'];
    final y = rawY is num && rawY.isFinite ? (rawY < 0 ? 0 : rawY.toInt()) : 0;
    final f = parsed['f'];
    return IReturnTicket(
      search: search,
      scrollY: y,
      focusId: f is String && f.isNotEmpty ? f : null,
    );
  } catch (_) {
    return null;
  }
}

/// 返回按钮上该写什么。带上「回到第几页」比光写「返回」有用得多
String returnLabel(IReturnTicket? ticket, {String fallback = '返回列表'}) {
  if (ticket == null) return fallback;
  final match = RegExp(r'(?:^|[?&])page=(\d+)').firstMatch(ticket.search);
  if (match == null) return fallback;
  // page 参数从 0 起算的居多，显示时补成人读的第几页
  return '返回列表第 ${int.parse(match.group(1)!) + 1} 页';
}

File: packages/common/src/logic/detail.ts
/**
 * 详情页的纯逻辑（astra.md 的 B12）。
 *
 * 详情页看着只是「把一条记录摊开」，真正难的是四件事，而且它们都不是排版问题：
 *
 * **一、这条记录现在能做什么。**
 * 同一张单，草稿态能编辑能提交，已发货只能打印。做法上有一条容易走反的分界：
 *
 *   状态不允许的动作  **不出现**。它在这个状态下根本不是一个动作，
 *                     灰着摆在那儿只会让人反复去试、去猜要满足什么条件。
 *   没权限的动作      **出现，但停用并说明原因**。藏起来的话，用户会以为
 *                     这个功能不存在，转头去问同事、提工单——而答案其实只是
 *                     「你没有这个权限」。
 *
 * **二、看到的这一份还作不作数。**
 * 详情页会停留很久。期间别人改了、删了，而屏幕上还是进来时那一份。
 * 这时候所有写动作都要停掉，并且明说「你看到的是 v3，现在已经是 v5」——
 * 只说「操作失败」等于让用户再点一次。
 *
 * **三、回到列表要回到原来那个位置。**
 * 从第 7 页翻到的那条记录，返回时落在第 1 页顶部，等于把用户之前的翻页与
 * 筛选全作废了。所以「从哪儿来」是一张要带着走的票据，不是靠浏览器后退去赌。
 *
 * **四、上一条 / 下一条。**
 * 详情页十有八九是逐条看过去的。到头了不要把按钮藏掉——藏掉会让人以为
 * 是页面坏了；说清「已经是第一条」。
 */

/* ---------- 这一份还作不作数 ---------- */

export type RecordFreshness = 'fresh' | 'stale' | 'deleted'

export interface FreshnessInput {
  /** 进页面时拿到的版本号 */
  seenRevision: number
  /** 此刻服务端上的版本号 */
  currentRevision: number
  /** 记录还在吗。删掉之后版本号无从比较 */
  exists?: boolean
}

export interface FreshnessState {
  kind: RecordFreshness
  /** 状态本身。颜色不是唯一线索，这句话必须出现 */
  label: string
  detail: string
  /** 这一刻该给的出口 */
  action: 'none' | 'refresh' | 'back'
}

export function recordFreshness(input: FreshnessInput): FreshnessState {
  if (input.exists === false) {
    return {
      kind: 'deleted',
      label: '已被删除',
      detail: '这条记录已经不在了，页面上显示的是你进来时的那一份',
      action: 'back'
    }
  }
  if (input.currentRevision > input.seenRevision) {
    return {
      kind: 'stale',
      label: '已被他人更新',
      // 只说「操作失败」等于让用户再点一次，所以把两个版本号都摆出来
      detail: `你看到的是 v${input.seenRevision}，现在已经是 v${input.currentRevision}`,
      action: 'refresh'
    }
  }
  return { kind: 'fresh', label: '最新', detail: '', action: 'none' }
}

/* ---------- 这条记录现在能做什么 ---------- */

export type DetailActionKind = 'primary' | 'default' | 'danger'

export interface DetailActionSpec {
  key: string
  label: string
  kind?: DetailActionKind
  /**
   * 这个动作在哪些状态下才算一个动作。不给表示任何状态都算。
   * 不在其中就**不出现**——不是灰着。
   */
  states?: readonly string[]
  /** 需要的权限。缺了就出现但停用，并说明原因 */
  permission?: string
  /**
   * 只读动作（导出、打印、复制链接）。记录失效时它们照样可用——
   * 把「打印」一起停掉，只会让用户以为整页坏了。
   */
  readonly?: boolean
}

export interface DetailActionContext {
  status: string
  permissions?: readonly string[]
  freshness?: RecordFreshness
  /**
   * 业务规则挡下的动作：键是动作 key，值是给人看的原因。
   * 例如「不能审批自己提交的单据」——这类规则写不进状态机也写不进权限表。
   */
  denied?: Readonly<Record<string, string>>
}

export interface DetailAction {
  key: string
  label: string
  kind: DetailActionKind
  disabled: boolean
  /** 停用原因。disabled 为 true 时一定不是空串 */
  reason: string
}

/**
 * 算出这一刻该摆出哪些动作、哪些是灰的、为什么灰。
 *
 * 三条停用理由的优先级：**权限 > 失效 > 业务规则**。
 * 权限排第一是因为它最稳定——刷新一百次也不会变，先说它才不会让用户
 * 去做无用功；失效排第二，它刷新一下就没了；业务规则最靠后，
 * 因为它通常是最具体、最长的一句话，前两条成立时说它反而模糊了重点。
 */
export function detailActions(
  specs: readonly DetailActionSpec[],
  context: DetailActionContext
): DetailAction[] {
  const permissions = new Set(context.permissions ?? [])
  const denied = context.denied ?? {}
  const out: DetailAction[] = []

  for (const spec of specs) {
    // 状态不允许的动作根本不出现：灰着摆在那儿只会让人反复去试
    if (spec.states && !spec.states.includes(context.status)) continue

    let reason = ''
    if (spec.permission && !permissions.has(spec.permission)) {
      reason = `需要「${spec.permission}」权限`
    } else if (!spec.readonly && context.freshness && context.freshness !== 'fresh') {
      reason =
        context.freshness === 'deleted'
          ? '这条记录已被删除'
          : '这条记录已被他人更新，请先刷新'
    } else if (denied[spec.key]) {
      reason = denied[spec.key]
    }

    out.push({
      key: spec.key,
      label: spec.label,
      kind: spec.kind ?? 'default',
      disabled: reason !== '',
      reason
    })
  }
  return out
}

/** 一个动作都没有时说什么。空白一片会让人以为页面没加载完 */
export function noActionHint(status: string): string {
  return `「${status}」状态下没有可执行的操作`
}

/* ---------- 上一条 / 下一条 ---------- */

export interface Neighbours {
  /** 在这批 id 里排第几，从 0 起。不在其中是 -1 */
  index: number
  total: number
  prevId: string | null
  nextId: string | null
  /** 「第 3 条，共 128 条」。不在这批里时是空串 */
  position: string
  /** 到头了说什么。没到头是空串 */
  edgeHint: string
}

export function detailNeighbours(ids: readonly string[], currentId: string): Neighbours {
  const index = ids.indexOf(currentId)
  const total = ids.length
  if (index < 0) {
    return { index: -1, total, prevId: null, nextId: null, position: '', edgeHint: '' }
  }
  const prevId = index > 0 ? ids[index - 1] : null
  const nextId = index < total - 1 ? ids[index + 1] : null
  // 到头了不把按钮藏掉——藏掉会让人以为是页面坏了
  const edgeHint =
    total <= 1 ? '只有这一条' : !prevId ? '已经是第一条' : !nextId ? '已经是最后一条' : ''
  return { index, total, prevId, nextId, position: `第 ${index + 1} 条，共 ${total} 条`, edgeHint }
}

/* ---------- 从哪儿来，回哪儿去 ---------- */

export interface ReturnTicket {
  /** 列表当时的查询串（含筛选与页码），原样带回去 */
  search: string
  /** 列表当时滚到哪儿。取整到像素——小数在不同缩放下还原不回同一位置 */
  scrollY: number
  /** 从哪一行点进来的。回去之后把它高亮一下，用户才知道自己停在哪 */
  focusId?: string
}

/**
 * 把「从哪儿来」打包成一段可以放进 URL 或 storage 的文本。
 *
 * 用 JSON 而不是自定义分隔符：查询串里本来就什么字符都可能有，
 * 自己拼分隔符迟早会被一个带 `&` 的筛选值劈开。
 */
export function packReturn(ticket: ReturnTicket): string {
  const payload: Record<string, unknown> = {
    s: ticket.search,
    y: Math.max(0, Math.round(ticket.scrollY))
  }
  if (ticket.focusId) payload.f = ticket.focusId
  return JSON.stringify(payload)
}

/**
 * 解回来。
 *
 * 解不出来就返回 null，不抛错：这张票据来自 URL，用户会手改、会截断、
 * 会从聊天软件里粘贴一个被吃掉尾巴的版本。为这种情况白屏是不可接受的，
 * 调用方拿到 null 就退回列表顶部——那是个不精确但绝不出错的落点。
 */
export function unpackReturn(raw: string | null | undefined): ReturnTicket | null {
  if (!raw) return null
  try {
    const parsed = JSON.parse(raw) as Record<string, unknown>
    if (typeof parsed !== 'object' || parsed === null) return null
    const search = typeof parsed.s === 'string' ? parsed.s : ''
    const y = typeof parsed.y === 'number' && Number.isFinite(parsed.y) ? Math.max(0, parsed.y) : 0
    const focusId = typeof parsed.f === 'string' && parsed.f ? parsed.f : undefined
    return { search, scrollY: y, focusId }
  } catch {
    return null
  }
}

/** 返回按钮上该写什么。带上「回到第几页」比光写「返回」有用得多 */
export function returnLabel(ticket: ReturnTicket | null, fallback = '返回列表'): string {
  if (!ticket) return fallback
  const page = /(?:^|[?&])page=(\d+)/.exec(ticket.search)
  if (!page) return fallback
  // page 参数从 0 起算的居多，显示时补成人读的第几页
  return `返回列表第 ${Number(page[1]) + 1} 页`
}
